rfcs/0002_amethyst_as_bin.md

```diff
@@ -0,0 +1,144 @@
+`!Default` resources.  Most commonly, these resources come from the rendering system, such as `ScreenDimensions`.  In order to
+gain the most benefits from this we are going to have to introduce a "core dispatcher" concept to the engine.
+The core dispatcher would be responsible for input, and rendering, both of which would be optional.  If neither are enabled
+then we simply don't run the core dispatcher.
+
+2. Editor. One could imagine the user crate as a pluggable set of code into a variety of possible amethyst executables.
+One of these executables could be the standalone engine, and another one could be a scene editor.  This would allow us to provide a
+"play" button that starts executing the user code, and since we control the flow of execution now we can also "stop" gameplay
+whenever we feel.
+
+3. Scripting. Part of our executable can be identifying, incorporating, and executing code written in foreign programming languages. 
+Since we control the flow of execution here we are no longer dependent on the user authoring Rust code in order to identify and
+execute their scripts. In theory, you could use Amethyst without needing to know Rust at all.
+
+4. Crossplatform compilation?  It's unclear if this benefit would be achievable as we don't have a proof of concept but it's possible
+that this model would allow us to compile a game out to Apple devices without actually owning an Apple device.  It's worth noting
+this feature only really matters to hobbyist users, as professionals are going to insist on QAing their build for which they'll still
+need actual player hardware.
+
+# Guide-Level Explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When generating a new Amethyst project you'll find a file in your project called `lib.rs`. In this file will be several
+functions Amethyst depends on you to provide.  Do not remove them, otherwise your project will not work with Amethyst.
+
+However, you can and are expected to alter the body of the function to change what is returned. Do not change the function names,
+parameters, or return types.
+
+Here are the functions and types we expect you to provide:
+
+- `pub type GameData`
+
+  This is the type used to store your system dispatchers and any thread local data you may have.  We've pre-populated it with a
+  simple to use type from Amethyst called `GameData` but you may want to provide your own type here later when your game gets
+  more complicated.
+  
+- `pub type Event`
+
+  This is the type of event that Amethyst will send to your states.  We've populated it with a default called `StateEvent`.
+
+- `pub fn first_state() -> impl State<GameData, Event>`
+  
+  Provides the first state that your game will execute. In the template file we've provided a basic empty state
+  and we are returning it in this function.
+  
+- `pub fn input_bindings() -> String`
```

Comment:
If you don't have input (i.e server), this will look weird.

Reply:
Which is why the option to return an empty string is provided.  It seems very non-trivial to have optional functions, so I am assuming every function we could possibly want is required, and we'll just provide a default implementation in the tooling.